File: cpp/examples/basic.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2022-2025, NVIDIA CORPORATION & AFFILIATES.
 * SPDX-License-Identifier: BSD-3-Clause
 */
#include <cassert>
#include <chrono>
#include <iostream>
#include <memory>
#include <numeric>
#include <string>
#include <thread>
#include <unistd.h>
#include <utility>
#include <vector>

#include <ucxx/api.h>
#include <ucxx/buffer.h>
#include <ucxx/utils/sockaddr.h>
#include <ucxx/utils/ucx.h>

#if UCXX_ENABLE_RMM
#include <rmm/device_buffer.hpp>
#endif

class ListenerContext {
 private:
  std::shared_ptr<ucxx::Worker> _worker{nullptr};
  std::shared_ptr<ucxx::Endpoint> _endpoint{nullptr};
  std::shared_ptr<ucxx::Listener> _listener{nullptr};

 public:
  explicit ListenerContext(std::shared_ptr<ucxx::Worker> worker) : _worker{worker} {}

  ~ListenerContext() { releaseEndpoint(); }

  void setListener(std::shared_ptr<ucxx::Listener> listener) { _listener = listener; }

  std::shared_ptr<ucxx::Listener> getListener() { return _listener; }

  std::shared_ptr<ucxx::Endpoint> getEndpoint() { return _endpoint; }

  bool isAvailable() const { return _endpoint == nullptr; }

  void createEndpointFromConnRequest(ucp_conn_request_h conn_request)
  {
    if (!isAvailable()) throw std::runtime_error("Listener context already has an endpoint");

    static bool endpoint_error_handling = true;
    _endpoint = _listener->createEndpointFromConnRequest(conn_request, endpoint_error_handling);
  }

  void releaseEndpoint() { _endpoint.reset(); }
};

static void listener_cb(ucp_conn_request_h conn_request, void* arg)
{
  char ip_str[INET6_ADDRSTRLEN];
  char port_str[INET6_ADDRSTRLEN];
  ucp_conn_request_attr_t attr{};
  ListenerContext* listener_ctx = reinterpret_cast<ListenerContext*>(arg);

  attr.field_mask = UCP_CONN_REQUEST_ATTR_FIELD_CLIENT_ADDR;
  ucxx::utils::ucsErrorThrow(ucp_conn_request_query(conn_request, &attr));
  ucxx::utils::sockaddr_get_ip_port_str(&attr.client_address, ip_str, port_str, INET6_ADDRSTRLEN);
  std::cout << "Server received a connection request from client at address " << ip_str << ":"
            << port_str << std::endl;

  if (listener_ctx->isAvailable()) {
    listener_ctx->createEndpointFromConnRequest(conn_request);
  } else {
    // The server is already handling a connection request from a client,
    // reject this new one
    std::cout << "Rejecting a connection request from " << ip_str << ":" << port_str << "."
              << std::endl
              << "Only one client at a time is supported." << std::endl;
    ucxx::utils::ucsErrorThrow(
      ucp_listener_reject(listener_ctx->getListener()->getHandle(), conn_request));
  }
}

static void printUsage()
{
  std::cerr << "Usage: basic [parameters]" << std::endl;
  std::cerr << " basic client/server example" << std::endl;
  std::cerr << std::endl;
  std::cerr << "Parameters are:" << std::endl;
  std::cerr << "  -m          progress mode to use, valid values are: 'polling', 'blocking',"
            << std::endl;
  std::cerr << "              'thread-polling', 'thread-blocking' and 'wait' (default: 'blocking')"
            << std::endl;
  std::cerr << "  -p <port>   Port number to listen at" << std::endl;
  std::cerr << "  -s <send_buffer_type>   Send buffer type, valid values are: 'host', 'rmm' "
               "(default: 'host')"
            << std::endl;
  std::cerr << "  -r <recv_buffer_type>   Recv buffer type, valid values are: 'host', 'rmm' "
               "(default: 'host')"
            << std::endl;
  std::cerr << "  -h          Print this help" << std::endl;
  std::cerr << std::endl;
}

enum class ProgressMode {
  Polling,
  Blocking,
  Wait,
  ThreadPolling,
  ThreadBlocking,
};

struct args {
  ProgressMode progress_mode{ProgressMode::Blocking};
  uint16_t listener_port{12345};
  ucxx::BufferType send_buf_type{ucxx::BufferType::Host};
  ucxx::BufferType recv_buf_type{ucxx::BufferType::Host};

  ucs_status_t parse(int argc, char* const argv[])
  {
    int c;

    auto parseBufferType = [](const std::string& bufferTypeString) {
      if (bufferTypeString == "host") {
        return ucxx::BufferType::Host;
      } else if (bufferTypeString == "rmm") {
#if UCXX_ENABLE_RMM
        return ucxx::BufferType::RMM;
#else
        std::cerr << "RMM support not enabled, please compile with -DUCXX_ENABLE_RMM=1"
                  << std::endl;
        return ucxx::BufferType::Invalid;
#endif
      } else {
        return ucxx::BufferType::Invalid;
      }
    };

    while ((c = getopt(argc, argv, "m:p:s:r:h")) != -1) {
      switch (c) {
        case 'm':
          if (strcmp(optarg, "blocking") == 0) {
            progress_mode = ProgressMode::Blocking;
            break;
          } else if (strcmp(optarg, "polling") == 0) {
            progress_mode = ProgressMode::Polling;
            break;
          } else if (strcmp(optarg, "thread-blocking") == 0) {
            progress_mode = ProgressMode::ThreadBlocking;
            break;
          } else if (strcmp(optarg, "thread-polling") == 0) {
            progress_mode = ProgressMode::ThreadPolling;
            break;
          } else if (strcmp(optarg, "wait") == 0) {
            progress_mode = ProgressMode::Wait;
            break;
          } else {
            std::cerr << "Invalid progress mode: " << optarg << std::endl;
            return UCS_ERR_INVALID_PARAM;
          }
        case 'p':
          listener_port = atoi(optarg);
          if (listener_port <= 0) {
            std::cerr << "Wrong listener port: " << listener_port << std::endl;
            return UCS_ERR_UNSUPPORTED;
          }
          break;
        case 's':
          send_buf_type = parseBufferType(optarg);
          if (send_buf_type == ucxx::BufferType::Invalid) {
            std::cerr << "Invalid send buffer type: " << optarg << std::endl;
            return UCS_ERR_INVALID_PARAM;
          }
          break;
        case 'r':
          recv_buf_type = parseBufferType(optarg);
          if (recv_buf_type == ucxx::BufferType::Invalid) {
            std::cerr << "Invalid recv buffer type: " << optarg << std::endl;
            return UCS_ERR_INVALID_PARAM;
          }
          break;
        case 'h':
        default: printUsage(); return UCS_ERR_UNSUPPORTED;
      }
    }

    return UCS_OK;
  }
};

std::function<void()> getProgressFunction(std::shared_ptr<ucxx::Worker> worker,
                                          ProgressMode progressMode)
{
  switch (progressMode) {
    case ProgressMode::Polling: return std::bind(std::mem_fn(&ucxx::Worker::progress), worker);
    case ProgressMode::Blocking:
      return std::bind(std::mem_fn(&ucxx::Worker::progressWorkerEvent), worker, -1);
    case ProgressMode::Wait: return std::bind(std::mem_fn(&ucxx::Worker::waitProgress), worker);
    default: return []() {};
  }
}

void waitRequests(ProgressMode progressMode,
                  std::shared_ptr<ucxx::Worker> worker,
                  const std::vector<std::shared_ptr<ucxx::Request>>& requests)
{
  auto progress = getProgressFunction(worker, progressMode);
  // Wait until all messages are completed
  for (auto& r : requests) {
    while (!r->isCompleted())
      progress();
    r->checkError();
  }
}

template <typename T>
std::shared_ptr<ucxx::Buffer> makeBuffer(ucxx::BufferType bufferType, T* values, size_t size)
{
  switch (bufferType) {
    case ucxx::BufferType::Host:
      return std::make_shared<ucxx::HostBuffer>(values, size * sizeof(T));
    case ucxx::BufferType::RMM:
#if UCXX_ENABLE_RMM
    {
      auto buf =
        std::make_unique<rmm::device_buffer>(values, size * sizeof(T), rmm::cuda_stream_default);
      rmm::cuda_stream_default.synchronize();
      return std::make_shared<ucxx::RMMBuffer>(std::move(buf));
    }
#endif
    default: throw std::runtime_error("Unable to make buffer from values");
  }
}

auto verify_buffers(ucxx::Buffer* expected, ucxx::Buffer* actual)
{
  std::vector<uint8_t> host_expected, host_actual;
  void *host_expected_ptr, *host_actual_ptr;

#if UCXX_ENABLE_RMM
  auto copy_to_host = [](auto& buffer, auto& host_buffer) {
    // copy RMM buffer to host
    host_buffer.resize(buffer->getSize());
    auto stream = rmm::cuda_stream_default;
    assert(
      cudaMemcpyAsync(
        host_buffer.data(), buffer->data(), buffer->getSize(), cudaMemcpyDefault, stream.value()) ==
      cudaSuccess);
    stream.synchronize();
    return host_buffer.data();
  };
#endif

  if (expected->getType() == ucxx::BufferType::RMM) {
#if UCXX_ENABLE_RMM
    host_expected_ptr = copy_to_host(expected, host_expected);
#else
    throw std::runtime_error("RMM support not enabled, please compile with -DUCXX_ENABLE_RMM=1");
#endif
  } else {
    host_expected_ptr = expected->data();
  }

  if (actual->getType() == ucxx::BufferType::RMM) {
#if UCXX_ENABLE_RMM
    host_actual_ptr = copy_to_host(actual, host_actual);
#else
    throw std::runtime_error("RMM support not enabled, please compile with -DUCXX_ENABLE_RMM=1");
#endif
  } else {
    host_actual_ptr = actual->data();
  }

  assert(std::memcmp(host_expected_ptr, host_actual_ptr, expected->getSize()) == 0);
}

int main(int argc, char** argv)
{
  args args;
  if (args.parse(argc, argv) != UCS_OK) return -1;

  // Setup: create UCP context, worker, listener and client endpoint.
  auto context      = ucxx::createContext({}, ucxx::Context::defaultFeatureFlags);
  auto worker       = context->createWorker();
  auto listener_ctx = std::make_unique<ListenerContext>(worker);
  auto listener     = worker->createListener(args.listener_port, listener_cb, listener_ctx.get());
  listener_ctx->setListener(listener);
  auto endpoint = worker->createEndpointFromHostname("127.0.0.1", args.listener_port, true);

  // Initialize worker progress
  if (args.progress_mode == ProgressMode::Blocking)
    worker->initBlockingProgressMode();
  else if (args.progress_mode == ProgressMode::ThreadBlocking)
    worker->startProgressThread(false);
  else if (args.progress_mode == ProgressMode::ThreadPolling)
    worker->startProgressThread(true);

  auto progress = getProgressFunction(worker, args.progress_mode);

  // Block until client connects
  while (listener_ctx->isAvailable())
    progress();

  std::vector<std::shared_ptr<ucxx::Request>> requests;

  // Allocate send buffers
  std::vector<int> wireupValues{1, 2, 3};
  auto sendWireupBuffer = makeBuffer(
    ucxx::BufferType::Host, wireupValues.data(), wireupValues.size());  // host wireup buffer

  std::vector<int> sendValues(50000);
  std::iota(sendValues.begin(), sendValues.end(), 0);
  std::vector<std::shared_ptr<ucxx::Buffer>> sendBuffers{
    makeBuffer(args.send_buf_type, sendValues.data(), 5),
    makeBuffer(args.send_buf_type, sendValues.data(), 500),
    makeBuffer(args.send_buf_type, sendValues.data(), 50000),
  };

  // Allocate receive buffers
  auto recvWireupBuffer =
    allocateBuffer(ucxx::BufferType::Host, sendWireupBuffer->getSize());  // host wireup buffer
  std::vector<std::shared_ptr<ucxx::Buffer>> recvBuffers;
  for (const auto& v : sendBuffers)
    recvBuffers.push_back(allocateBuffer(args.recv_buf_type, v->getSize()));

  // Schedule small wireup messages to let UCX identify capabilities between endpoints
  requests.push_back(listener_ctx->getEndpoint()->tagSend(
    sendWireupBuffer->data(), sendWireupBuffer->getSize(), ucxx::Tag{0}));
  requests.push_back(endpoint->tagRecv(
    recvWireupBuffer->data(), sendWireupBuffer->getSize(), ucxx::Tag{0}, ucxx::TagMaskFull));
  ::waitRequests(args.progress_mode, worker, requests);
  requests.clear();

  // Schedule send and recv messages on different tags and different ordering
  requests.push_back(listener_ctx->getEndpoint()->tagSend(
    sendBuffers[0]->data(), sendBuffers[0]->getSize(), ucxx::Tag{0}));
  requests.push_back(listener_ctx->getEndpoint()->tagRecv(
    recvBuffers[1]->data(), recvBuffers[1]->getSize(), ucxx::Tag{1}, ucxx::TagMaskFull));
  requests.push_back(listener_ctx->getEndpoint()->tagSend(
    sendBuffers[2]->data(), sendBuffers[2]->getSize(), ucxx::Tag{2}, ucxx::TagMaskFull));
  requests.push_back(endpoint->tagRecv(
    recvBuffers[2]->data(), recvBuffers[2]->getSize(), ucxx::Tag{2}, ucxx::TagMaskFull));
  requests.push_back(
    endpoint->tagSend(sendBuffers[1]->data(), sendBuffers[1]->getSize(), ucxx::Tag{1}));
  requests.push_back(endpoint->tagRecv(
    recvBuffers[0]->data(), recvBuffers[0]->getSize(), ucxx::Tag{0}, ucxx::TagMaskFull));

  // Wait for requests to be set, i.e., transfers complete
  ::waitRequests(args.progress_mode, worker, requests);

  // Verify results
  verify_buffers(sendWireupBuffer.get(), recvWireupBuffer.get());
  for (size_t i = 0; i < sendBuffers.size(); ++i) {
    verify_buffers(sendBuffers[i].get(), recvBuffers[i].get());
  }

  // Stop progress thread
  if (args.progress_mode == ProgressMode::ThreadBlocking ||
      args.progress_mode == ProgressMode::ThreadPolling)
    worker->stopProgressThread();

  std::cout << "Example completed successfully" << std::endl;

  return 0;
}

File: python/examples/basic.py
# SPDX-FileCopyrightText: Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: BSD-3-Clause

import argparse
import asyncio

import numpy as np
import ucxx._lib.libucxx as ucx_api
from ucxx._lib.arr import Array
from ucxx._lib_async.utils import get_event_loop


def _create_cuda_context():
    import numba.cuda

    numba.cuda.current_context()


async def _progress_coroutine(worker):
    while True:
        try:
            if worker is None:
                return
            worker.progress()
            await asyncio.sleep(0)
        except asyncio.CancelledError:
            return


async def _wait_requests_async_future(loop, worker, requests):
    progress_task = loop.create_task(_progress_coroutine(worker))

    await asyncio.gather(*[r.future for r in requests])

    progress_task.cancel()


async def _wait_requests_async_yield(loop, worker, requests):
    progress_task = loop.create_task(_progress_coroutine(worker))

    await asyncio.gather(*[r.wait_yield() for r in requests])

    progress_task.cancel()


def _wait_requests(worker, progress_mode, requests):
    while not all([r.completed for r in requests]):
        if progress_mode == "blocking":
            worker.progress_worker_event()


def parse_args():
    parser = argparse.ArgumentParser(description="Basic UCXX-Py Example")
    parser.add_argument(
        "--asyncio-wait-future",
        default=False,
        action="store_true",
        help="Wait for transfer requests with Python's asyncio using futures "
        "(`UCXRequest.future`), requires `--progress-mode blocking`. "
        "(default: disabled)",
    )
    parser.add_argument(
        "--asyncio-wait-yield",
        default=False,
        action="store_true",
        help="Wait for transfer requests with Python's asyncio by checking "
        "for request completion and yielding (`UCXRequest.wait_yield()`). "
        "(default: disabled)",
    )
    parser.add_argument(
        "-m",
        "--progress-mode",
        default="thread",
        help="Progress mode for the UCP worker. Valid options are: "
        "'thread' (default) and 'blocking'.",
        type=str,
    )
    parser.add_argument(
        "-o",
        "--object-type",
        default="numpy",
        choices=["numpy", "rmm"],
        help="In-memory array type.",
        type=str,
    )
    parser.add_argument(
        "--multi-buffer-transfer",
        default=False,
        action="store_true",
        help="If specified, use the multi-buffer TAG transfer API.",
    )
    parser.add_argument(
        "-p",
        "--port",
        default=12345,
        help="The port the listener will bind to.",
        type=int,
    )

    args = parser.parse_args()
    valid_progress_modes = ["blocking", "thread"]
    if not any(args.progress_mode == v for v in valid_progress_modes):
        raise ValueError(
            f"Unknown progress mode '{args.progress_mode}', "
            f"valid modes are {valid_progress_modes}",
        )
    if args.asyncio_wait_future and args.progress_mode != "blocking":
        raise RuntimeError(
            "`--asyncio-wait-future` requires `--progress-mode='blocking'`"
        )

    return args


def main():
    args = parse_args()

    if args.object_type == "rmm":
        import cupy as xp

        import rmm
        from rmm.allocators.cupy import rmm_cupy_allocator

        rmm.reinitialize(
            pool_allocator=True,
            managed_memory=False,
        )
        xp.cuda.runtime.setDevice(0)
        xp.cuda.set_allocator(rmm_cupy_allocator)
    else:
        import numpy as xp

    ctx = ucx_api.UCXContext()

    worker = ucx_api.UCXWorker(ctx)

    if args.progress_mode == "blocking":
        worker.init_blocking_progress_mode()
    else:
        if args.object_type == "rmm":
            worker.set_progress_thread_start_callback(_create_cuda_context)

        worker.start_progress_thread()

    wireup_send_buf = np.arange(3)
    wireup_recv_buf = np.empty_like(wireup_send_buf)
    send_bufs = [
        xp.arange(50, dtype="u1"),
        xp.arange(500, dtype="u1"),
        xp.arange(50000, dtype="u1"),
    ]

    if args.multi_buffer_transfer is False:
        recv_bufs = [np.empty_like(b) for b in send_bufs]

    global listener_ep
    listener_ep = None

    def listener_callback(conn_request):
        global listener_ep
        listener_ep = listener.create_endpoint_from_conn_request(conn_request, True)

    listener = ucx_api.UCXListener.create(
        worker,
        args.port,
        listener_callback,
    )

    ep = ucx_api.UCXEndpoint.create(
        worker,
        "127.0.0.1",
        args.port,
        endpoint_error_handling=True,
    )

    while listener_ep is None:
        if args.progress_mode == "blocking":
            worker.progress_worker_event()

    wireup_requests = [
        ep.tag_send(Array(wireup_send_buf), tag=ucx_api.UCXXTag(0)),
        listener_ep.tag_recv(Array(wireup_recv_buf), tag=ucx_api.UCXXTag(0)),
    ]
    _wait_requests(worker, args.progress_mode, wireup_requests)

    np.testing.assert_equal(wireup_recv_buf, wireup_send_buf)

    if args.multi_buffer_transfer:
        frames = tuple(
            [
                Array(send_bufs[0]),
                Array(send_bufs[1]),
                Array(send_bufs[2]),
            ]
        )

        # data_ptrs = tuple(f.ptr for f in frames)
        # sizes = tuple(f.nbytes for f in frames)
        # is_cuda = tuple(f.cuda for f in frames)
        # send_buffer_requests = listener_ep.tag_send_multi(
        #     data_ptrs, sizes, is_cuda, tag=0
        # )

        send_buffer_requests = listener_ep.tag_send_multi(
            frames, tag=ucx_api.UCXXTag(0)
        )
        recv_buffer_requests = ep.tag_recv_multi(0)

        requests = [send_buffer_requests, recv_buffer_requests]

        if args.asyncio_wait_future:
            loop = get_event_loop()
            loop.run_until_complete(_wait_requests_async_future(loop, worker, requests))
        elif args.asyncio_wait_yield:
            loop = get_event_loop()
            loop.run_until_complete(_wait_requests_async_yield(loop, worker, requests))
        else:
            _wait_requests(worker, args.progress_mode, requests)

            # Check results, raises an exception if any of them failed
            for r in (
                send_buffer_requests.get_requests()
                + recv_buffer_requests.get_requests()
            ):
                r.check_error()

        recv_bufs = recv_buffer_requests.get_py_buffers()
    else:
        requests = [
            listener_ep.tag_send(Array(send_bufs[0]), tag=ucx_api.UCXTag(0)),
            listener_ep.tag_send(Array(send_bufs[1]), tag=ucx_api.UCXTag(1)),
            listener_ep.tag_send(Array(send_bufs[2]), tag=ucx_api.UCXTag(2)),
            ep.tag_recv(Array(recv_bufs[0]), tag=ucx_api.UCXTag(0)),
            ep.tag_recv(Array(recv_bufs[1]), tag=ucx_api.UCXTag(1)),
            ep.tag_recv(Array(recv_bufs[2]), tag=ucx_api.UCXTag(2)),
        ]

        if args.asyncio_wait_future:
            loop = get_event_loop()
            loop.run_until_complete(_wait_requests_async_future(loop, worker, requests))
        elif args.asyncio_wait_yield:
            loop = get_event_loop()
            loop.run_until_complete(_wait_requests_async_yield(loop, worker, requests))
        else:
            _wait_requests(worker, args.progress_mode, requests)

            # Check results, raises an exception if any of them failed
            for r in requests:
                r.check_error()

    if args.progress_mode == "thread":
        worker.stop_progress_thread()

    for recv_buf, send_buf in zip(recv_bufs, send_bufs):
        if args.object_type == "numpy":
            xp.testing.assert_equal(recv_buf, send_buf)
        else:
            xp.testing.assert_array_equal(xp.asarray(recv_buf), send_buf)


if __name__ == "__main__":
    main()
